File: examples/shared-hosting/python/loki_handler.py
"""
LokiHandler - Python logging handler that pushes logs to Loki HTTP API.

For shared hosting or environments where you can't install Promtail.

Usage:
    import logging
    from loki_handler import LokiHandler

    logger = logging.getLogger('myapp')
    handler = LokiHandler(
        url='https://YOUR_VPS:3101',
        username='pantra-push',
        password='your_strong_password',
        labels={'app': 'myapp', 'env': 'production'},
    )
    logger.addHandler(handler)
    logger.setLevel(logging.INFO)

    logger.info('server started on port 8000')
    logger.error('DB connection failed', exc_info=True)

@see https://github.com/cahrur/pantra/blob/main/SHARED-HOSTING.md
"""

import json
import time
import threading
import logging
from urllib.request import Request, urlopen
from urllib.error import URLError
from base64 import b64encode


class LokiHandler(logging.Handler):
    """Push logs to Loki HTTP API. Works on shared hosting (no binary needed)."""

    def __init__(self, url, username='', password='', labels=None,
                 batch_size=10, flush_interval=5.0, level=logging.DEBUG):
        super().__init__(level)
        self.url = url.rstrip('/') + '/loki/api/v1/push'
        self.auth = None
        if username and password:
            cred = b64encode(f'{username}:{password}'.encode()).decode()
            self.auth = f'Basic {cred}'
        self.labels = labels or {}
        self.batch_size = batch_size
        self.buffer = []
        self.lock = threading.Lock()

        # Auto-flush timer
        self._flush_interval = flush_interval
        self._timer = threading.Timer(flush_interval, self._auto_flush)
        self._timer.daemon = True
        self._timer.start()

    def emit(self, record):
        with self.lock:
            self.buffer.append(record)
            if len(self.buffer) >= self.batch_size:
                self._flush()

    def _auto_flush(self):
        with self.lock:
            self._flush()
        self._timer = threading.Timer(self._flush_interval, self._auto_flush)
        self._timer.daemon = True
        self._timer.start()

    def _flush(self):
        if not self.buffer:
            return
        entries = self.buffer[:]
        self.buffer = []

        # Group by label set for better Loki performance
        grouped = {}
        for record in entries:
            labels = {**self.labels, 'level': record.levelname.lower()}
            key = json.dumps(labels, sort_keys=True)
            if key not in grouped:
                grouped[key] = {'stream': labels, 'values': []}
            ts = str(int(record.created * 1e9))
            msg = self.format(record) if self.formatter else record.getMessage()
            grouped[key]['values'].append([ts, msg])

        payload = json.dumps({'streams': list(grouped.values())}).encode()
        req = Request(self.url, data=payload, method='POST')
        req.add_header('Content-Type', 'application/json')
        if self.auth:
            req.add_header('Authorization', self.auth)

        try:
            urlopen(req, timeout=3)
        except (URLError, OSError):
            pass  # fire and forget

    def close(self):
        self._timer.cancel()
        with self.lock:
            self._flush()
        super().close()

File: examples/shared-hosting/php/LokiHandler.php
<?php
/**
 * LokiHandler - Monolog handler that pushes logs to Loki HTTP API.
 * 
 * For shared hosting environments where you can't install Promtail.
 * Copy this file to your Laravel app: app/Logging/LokiHandler.php
 *
 * @see https://github.com/cahrur/pantra/blob/main/SHARED-HOSTING.md
 */

namespace App\Logging;

use Monolog\Handler\AbstractProcessingHandler;
use Monolog\Level;
use Monolog\LogRecord;

class LokiHandler extends AbstractProcessingHandler
{
    private string $lokiUrl;
    private string $username;
    private string $password;
    private array $labels;
    private array $buffer = [];
    private int $batchSize;

    public function __construct(
        string $lokiUrl,
        string $username = '',
        string $password = '',
        array $labels = [],
        int $batchSize = 10,
        Level $level = Level::Debug,
        bool $bubble = true
    ) {
        parent::__construct($level, $bubble);
        $this->lokiUrl = rtrim($lokiUrl, '/') . '/loki/api/v1/push';
        $this->username = $username;
        $this->password = $password;
        $this->labels = $labels;
        $this->batchSize = $batchSize;

        register_shutdown_function([$this, 'flush']);
    }

    protected function write(LogRecord $record): void
    {
        $this->buffer[] = $record;

        if (count($this->buffer) >= $this->batchSize) {
            $this->flush();
        }
    }

    public function flush(): void
    {
        if (empty($this->buffer)) {
            return;
        }

        // Group entries by label set for better Loki performance
        $grouped = [];
        foreach ($this->buffer as $record) {
            $labels = array_merge($this->labels, [
                'level' => strtolower($record->level->name),
                'channel' => $record->channel,
            ]);

            $message = $record->formatted ?? $record->message;
            if (!empty($record->context)) {
                $message .= ' ' . json_encode($record->context);
            }

            $key = json_encode($labels);
            if (!isset($grouped[$key])) {
                $grouped[$key] = ['stream' => $labels, 'values' => []];
            }
            $grouped[$key]['values'][] = [
                (string)(intval($record->datetime->format('U.u') * 1e9)),
                $message,
            ];
        }

        $payload = json_encode(['streams' => array_values($grouped)]);
        $this->send($payload);
        $this->buffer = [];
    }

    private function send(string $payload): void
    {
        $ch = curl_init($this->lokiUrl);
        $headers = ['Content-Type: application/json'];

        curl_setopt_array($ch, [
            CURLOPT_POST => true,
            CURLOPT_POSTFIELDS => $payload,
            CURLOPT_HTTPHEADER => $headers,
            CURLOPT_TIMEOUT => 3,
            CURLOPT_CONNECTTIMEOUT => 2,
            CURLOPT_RETURNTRANSFER => true,
        ]);

        if ($this->username && $this->password) {
            curl_setopt($ch, CURLOPT_USERPWD, $this->username . ':' . $this->password);
        }

        curl_exec($ch);
        curl_close($ch);
    }
}
